### PR TITLE
🏅Audit IOSIRO

### PR DESCRIPTION
**Pull request by JaredBorders**
Please refer to the following notion page for audit response details:

https://asphyxia-grok.notion.site/Zap-Audit-10dfdfc841818017881ae5d21ab985d1?pvs=4

### CODE REVIEW

File: src/Zap.sol [hunk 1]
```diff
@@ -386,83 +428,108 @@ contract Zap is Errors {
...
+        /// @custom:synthetix debt is denominated in USDx
+        /// @custom:aave debt is denominated in USDC
+        /// @dev scale loan samount accordingly
+        amount /= 10 ** (usdxDecimals - usdcDecimals);
+
+        /// @dev barring exceptional circumstances,
+        /// a 1 USD buffer is sufficient to circumvent
+        /// precision loss
+        amount += 10 ** usdcDecimals;
     }
```
Comment (barrasso) [inline src/Zap.sol:547]: math checks out 👍

File: src/Zap.sol [hunk 2]
```diff
@@ -386,83 +428,108 @@ contract Zap is Errors {
...
+        amount = IPerpsMarket(PERPS_MARKET).debt(_accountId);
+
+        uint256 usdxDecimals = IERC20(USDX).decimals();
+        uint256 usdcDecimals = IERC20(USDC).decimals();
 
-        // establish amount of unwound collateral after deduction
-        unwound = unwrapped - deducted;
+        /// @custom:synthetix debt is denominated in USDx
+        /// @custom:aave debt is denominated in USDC
+        /// @dev scale loan samount accordingly
```
Comment (barrasso) [inline src/Zap.sol]: [np] minor typo `samount`

File: src/Zap.sol [hunk 3]
```diff
@@ -370,13 +409,16 @@ contract Zap is Errors {
             params: params,
             referralCode: REFERRAL_CODE
         });
+
+        stage = Stage.UNSET;
```
Comment (barrasso) [inline src/Zap.sol:399]: won't `stage` be `UNSET` (and not `LEVEL1`) at execution time?

File: src/Zap.sol [hunk 4]
```diff
@@ -343,25 +369,38 @@ contract Zap is Errors {
...
         uint128 _collateralId,
+        uint256 _collateralAmount,
         uint256 _zapTolerance,
+        uint256 _unwrapTolerance,
         uint256 _swapTolerance,
         address _receiver
     )
         external
+        isAuthorized(_accountId)
+        requireStage(Stage.UNSET)
```
Comment (barrasso) [inline src/Zap.sol:372]: is it worth using the traditional `nonReentrant` modifier for external funcs in addition to the multi-stage guard, or is that overkill?

File: src/Zap.sol [hunk 5]
```diff
@@ -386,83 +428,108 @@ contract Zap is Errors {
...
-            _swapTolerance
+        (,,,,,, address _receiver) = abi.decode(
+            _params,
+            (uint128, uint128, uint256, uint256, uint256, uint256, address)
         );
 
-        _flashloan += _premium;
+        (uint256 unwound, address collateral) =
+            _unwind(_flashloan, _premium, _params);
+
```
Comment (barrasso) [inline src/Zap.sol:439]: is it worth emitting an event here for easier tracking of values / improved transparency?

(e.g. `Unwound(...)`)

File: src/Zap.sol [hunk 6]
```diff
@@ -485,7 +552,7 @@ contract Zap is Errors {
     /// @dev allowance is assumed
     /// @dev following execution, this contract will hold any excess USDx
     function _burn(uint256 _amount, uint128 _accountId) internal {
```
Comment (Flocqst) [inline src/Zap.sol:580]: When calling `burn`externally, if debt is repaid without overpaying, `remaining`will be `0`and `_push(USDX, msg.sender, remaining)` will revert as USDX does not support 0 amount transfers.

Need to check that `remaining > 0` before pushing back any excess USDx to the caller.

**Review by Flocqst — COMMENTED** [review]
Changes look good ! Left you a comment on `burn` when transferring back excess USDx.

File: src/Zap.sol (hunk 2 above)
Comment (JaredBorders) [reply src/Zap.sol]: fixed

File: src/Zap.sol (hunk 1 above)
Comment (JaredBorders) [reply src/Zap.sol:547]: ❤️

File: src/Zap.sol (hunk 6 above)
Comment (JaredBorders) [reply src/Zap.sol:580]: fixed good catch 🫡

File: src/Zap.sol (hunk 3 above)
Comment (JaredBorders) [reply src/Zap.sol:399]: Here is the breakdown of context and state during execution

1. `unwind()` confirms the `stage` is `UNSET` and sets it to `LEVEL1`
2. `flashLoanSimple()` is called, which itself will call (i.e., callback) `executeOperation()`, thus passing context to `executeOperation()`
3. `executeOperation()` confirms the `stage` is `LEVEL1` and sets it to `LEVEL2`
4. After `executeOperation()` returns context to `flashLoanSimple()`, it returns context to `unwind()`, which finally sets stage back to `UNSET`

Comment (JaredBorders) [reply src/Zap.sol:399]: e2e test will confirm this though

File: src/Zap.sol (hunk 4 above)
Comment (JaredBorders) [reply src/Zap.sol:372]: i think it is overkill. tbh I think even the multi-stage thing is a bit overkill but I can live with it

File: src/Zap.sol (hunk 5 above)
Comment (JaredBorders) [reply src/Zap.sol:439]: after audit I normally talk with data people to see what they need! they know best lol

**Review by Flocqst — APPROVED** [review]

**Review by moss-eth — APPROVED** [review]
LGTM

**Review by cmontecoding — APPROVED** [review]
LGTM

**Review by barrasso — APPROVED** [review]

**Pull request merged**